File: classical_elixir/btree_bfs/lib/btree_bfs/queue.ex
defmodule BtreeBfs.Queue do
  @moduledoc """
  队列的elixir实现
  """

  def insert(queue, nil), do: queue
  def insert(queue, item), do: :queue.in(item, queue)

  def insert_last(queue, nil), do: queue
  def insert_last(queue, item), do: :queue.in_r(item, queue)

  def member?(queue, item), do: :queue.member(item, queue)

  def filter(queue, fun), do: :queue.filter(fun, queue)

  def split(queue, n) do
    length = BtreeBfs.Queue.length(queue)

    if length >= n do
      :queue.split(n, queue)
    else
      :queue.split(length, queue)
    end
  end

  def take(queue, n) do
    {q1, _} = split(queue, n)
    q1
  end

  def pop(queue) do
    case :queue.out(queue) do
      {{:value, item}, new_queue} -> {item, new_queue}
      {:empty, new_queue} -> {nil, new_queue}
    end
  end

  def pop_last(queue) do
    case :queue.out_r(queue) do
      {{:value, item}, new_queue} -> {item, new_queue}
      {:empty, new_queue} -> {nil, new_queue}
    end
  end

  def first(queue) do
    unless is_empty?(queue) do
      :queue.get(queue)
    end
  end

  def last(queue) do
    unless is_empty?(queue) do
      :queue.get_r(queue)
    end
  end

  def head(queue) do
    unless is_empty?(queue) do
      :queue.head(queue)
    end
  end

  def tail(queue) do
    if is_empty?(queue) do
      queue
    else
      :queue.tail(queue)
    end
  end

  def drop(queue) do
    if is_empty?(queue) do
      queue
    else
      :queue.drop(queue)
    end
  end

  def drop_last(queue) do
    if is_empty?(queue) do
      queue
    else
      :queue.drop_r(queue)
    end
  end

  defdelegate length(queue), to: :queue, as: :len
  defdelegate from_list(queue), to: :queue
  defdelegate to_list(queue), to: :queue
  defdelegate is_empty?(queue), to: :queue, as: :is_empty
  defdelegate is_queue?(queue), to: :queue, as: :is_queue
  defdelegate join(queue1, queue2), to: :queue
  defdelegate reverse(queue), to: :queue
  defdelegate new(), to: :queue
end

File: classical_golang/queue/queue.go
package main

import (
	"fmt"
)

type node struct {
	value interface{}
	next  *node
}

type Queue struct {
	head *node
	tail *node
	size int
}

func (q *Queue) Size() int { return q.size }
func (q *Queue) Pop() interface{} {
	if q.head == nil {
		panic("empty queue")
	}

	var ans *node = q.head

	q.head = q.head.next
	q.size--
	if q.head == nil {
		q.tail = nil
	}

	return ans.value
}
func (q *Queue) Push(value interface{}) {
	var new_node *node = new(node)
	new_node.value = value
	new_node.next = nil

	if q.tail == nil {
		q.head = new_node
		q.tail = new_node
	} else {
		q.tail.next = new_node
		q.tail = new_node
	}
	q.size++
}

func main() {

	type val struct {
		x int
		y int
	}

	var q *Queue = new(Queue)
	var t *val = new(val)

	t.x = 1
	t.y = 2
	q.Push(*t)

	t.x = 3
	t.y = 4

	q.Push(*t)

	*t = q.Pop().(val)
	fmt.Println(*t)
	fmt.Println(q.Pop())
}
